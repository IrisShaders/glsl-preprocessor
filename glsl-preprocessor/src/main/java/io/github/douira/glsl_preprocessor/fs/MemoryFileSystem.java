package io.github.douira.glsl_preprocessor.fs;

import java.io.IOException;
import java.util.*;

public class MemoryFileSystem implements VirtualFileSystem {
	private final Map<String, VirtualFile> files = new HashMap<String, VirtualFile>();

	public MemoryFileSystem() {
	}

	public MemoryFileSystem(Map<String, VirtualFile> files) {
		this.files.putAll(files);
	}

	public void addFile(String path, VirtualFile file) {
		files.put(path, file);
	}

	public void addFile(String path, String content) {
		addFile(path, new StringFile(path, content));
	}

	@Override
	public VirtualFile getFile(String path) {
		if (!files.containsKey(path)) {
			throw new RuntimeException(new IOException("File not found: " + path));
		}
		return files.get(path);
	}
}
