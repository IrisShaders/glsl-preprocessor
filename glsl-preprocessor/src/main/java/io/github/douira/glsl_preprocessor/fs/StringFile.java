package io.github.douira.glsl_preprocessor.fs;

import edu.umd.cs.findbugs.annotations.NonNull;
import io.github.douira.glsl_preprocessor.*;

public class StringFile implements VirtualFile {
	private final String name;
	private final String content;

	public StringFile(String name, String content) {
		this.name = name;
		this.content = content;
	}

	@Override
	public boolean isFile() {
		return true;
	}

	@NonNull
	@Override
	public Source getSource() {
		return new StringLexerSource(content, name, true);
	}

	@Override
	public String toString() {
		return name;
	}
}
