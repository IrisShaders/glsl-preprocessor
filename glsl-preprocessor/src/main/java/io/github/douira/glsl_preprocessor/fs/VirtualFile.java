/*
 * Anarres C Preprocessor
 * Copyright (c) 2007-2015, Shevek
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied.  See the License for the specific language governing
 * permissions and limitations under the License.
 *
 * Modified by the contributors of glsl-preprocessor.
 */
package io.github.douira.glsl_preprocessor.fs;

import edu.umd.cs.findbugs.annotations.NonNull;
import io.github.douira.glsl_preprocessor.Source;

/**
 * An extremely lightweight virtual file interface.
 */
public interface VirtualFile {
	boolean isFile();

	@NonNull
	Source getSource();
}
