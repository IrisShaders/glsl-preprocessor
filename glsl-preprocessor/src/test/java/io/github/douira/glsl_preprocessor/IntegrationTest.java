package io.github.douira.glsl_preprocessor;

import au.com.origin.snapshots.Expect;
import au.com.origin.snapshots.annotations.SnapshotName;
import au.com.origin.snapshots.junit5.SnapshotExtension;
import edu.umd.cs.findbugs.annotations.NonNull;
import edu.umd.cs.findbugs.annotations.Nullable;
import io.github.douira.glsl_preprocessor.fs.MemoryFileSystem;
import io.github.douira.glsl_preprocessor.fs.StringFile;
import io.github.douira.glsl_preprocessor.fs.VirtualFile;
import io.github.douira.glsl_preprocessor.fs.VirtualFileSystem;
import io.github.douira.glsl_preprocessor.test_util.SnapshotUtil;
import io.github.douira.glsl_preprocessor.test_util.TestCaseProvider.Spacing;
import io.github.douira.glsl_preprocessor.test_util.TestCaseSource;
import org.junit.jupiter.api.extension.ExtendWith;
import org.junit.jupiter.params.ParameterizedTest;

import java.util.function.Consumer;

import static io.github.douira.glsl_preprocessor.test_util.AssertUtil.getBase64Hash;
import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertThrows;

@ExtendWith({SnapshotExtension.class})
public class IntegrationTest {
	private Expect expect;

	private void preprocessorTestCase(String type, String input, String expectedOutput, @Nullable Consumer<Preprocessor> prepare) {
		Preprocessor pp = new Preprocessor(input);
		var errors = false;
		for (var featureName : type.split(",")) {
			if (featureName.equals("errors")) {
				errors = true;
				continue;
			}
			var feature = Feature.valueOf(featureName);
			pp.addFeature(feature);
		}

		if (prepare != null) {
			prepare.accept(pp);
		}

		var snapshotOutput = new StringBuilder();
		if (errors) {
			var e = assertThrows(Exception.class, pp::printToString);
			pp.close();
			snapshotOutput.append(e.toString());
		} else {
			var output = pp.printToString();
			assertEquals(expectedOutput, output);
			snapshotOutput.append("(no error)");
		}
		
		for (var entry : pp.getSourceNumbers().entrySet()) {
			var source = entry.getKey();
			if (source == null) {
				continue;
			}
			var sourceNumber = entry.getValue();
			snapshotOutput.append("\n");
			snapshotOutput.append(sourceNumber);
			snapshotOutput.append("->");
			snapshotOutput.append(source);
		}

		expect
				.scenario(type + "_" + getBase64Hash(input))
				.toMatchSnapshot(SnapshotUtil.inputOutputSnapshot(
						input, snapshotOutput.toString()));
	}

	@SnapshotName("testPreprocessor")
	@ParameterizedTest
	@TestCaseSource(caseSet = "testPreprocessor", spacing = Spacing.TRIM_SINGLE_BOTH)
	void testPreprocessor(String type, String input, String expectedOutput) {
		preprocessorTestCase(type, input, expectedOutput, null);
	}

	private static class SyntheticLengthContentFileSystem implements VirtualFileSystem {
		@NonNull
		@Override
		public VirtualFile getFile(@NonNull String path) {
			var prefix = "length!";
			if (!path.startsWith(prefix)) {
				throw new IllegalArgumentException("File not found: " + path);
			}
			var command = path.substring(prefix.length()).split(":");
			var lengthTarget = Integer.parseInt(command[0]);
			var builder = new StringBuilder();
			for (int i = 0; i < lengthTarget; i++) {
				builder.append("content ");
				builder.append(i);
				builder.append('\n');
			}
			return new StringFile(path, builder.toString());
		}
	}

	@SnapshotName("testIncludeLineMapping")
	@ParameterizedTest
	@TestCaseSource(caseSet = "testIncludeLineMapping", spacing = Spacing.TRIM_SINGLE_BOTH)
	void testIncludeLineMapping(String type, String input, String expectedOutput) {
		preprocessorTestCase(type, input, expectedOutput, pp -> {
			pp.setFileSystem(new SyntheticLengthContentFileSystem());
		});
	}

	@SnapshotName("testIncludeLineMappingExplicitFiles")
	@ParameterizedTest
	@TestCaseSource(caseSet = "testIncludeLineMappingExplicitFiles", spacing = Spacing.TRIM_SINGLE_BOTH)
	void testIncludeLineMappingExplicitFiles(String type, String input, String expectedOutput) {
		var segments = input.split("@");
		var fs = new MemoryFileSystem();
		String ppInput = null;
		for (var segment : segments) {
			if (segment.isEmpty()) {
				continue;
			}
			var parts = segment.split("\n", 2);
			if (parts.length < 2) {
				throw new IllegalArgumentException("Invalid input: " + segment);
			}
			var path = parts[0].trim();
			var content = parts[1];
			if (path.equals("main")) {
				ppInput = content;
			} else {
				fs.addFile(path, content);
			}
		}
		preprocessorTestCase(type, ppInput, expectedOutput, pp -> {
			pp.setFileSystem(fs);
		});
	}
}
