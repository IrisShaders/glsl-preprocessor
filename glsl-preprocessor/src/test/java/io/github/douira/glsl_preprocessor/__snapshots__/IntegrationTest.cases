§testPreprocessor GLSL_PASSTHROUGH:
#version 330 core
foo bar
#extension foo : bar
#define DEFINED 4
int a = DEFINED;

===
#version 330 core
foo bar
#extension foo : bar

int a = 4;

§testPreprocessor GLSL_PASSTHROUGH:
#extension GL_ARB_texture_rectangle : enable
#extension GL_ARB_texture_rectangle : enable
#version 330 fjdsal fjsda fisda jfiojfdsaflkds
#line 1
#if 0
#endif
#line 2
int a = 4;
#define BAR 4
#define FOO 4
if (a == BAR) {
	a = FOO;
}

===
#extension GL_ARB_texture_rectangle : enable
#extension GL_ARB_texture_rectangle : enable
#version 330 fjdsal fjsda fisda jfiojfdsaflkds




int a = 4;


if (a == 4) {
	a = 4;
}

§testPreprocessor GLSL_PASSTHROUGH:
#line 5 324
#version 330 core
int a = 4;
#line 43 453

foo bar 
#ifdef FOO
FOO is defined here
#endif
#define FOO 5
foo bar
#ifdef FOO
FOO is defined here
#endif

===

#version 330 core
int a = 4;


foo bar 




foo bar

5 is defined here


§testPreprocessor GLSL_PASSTHROUGH,GLSL_CUSTOM_PASSTHROUGH,ARBITRARY_PRAGMAS:
#custom fdsaflsj 43lk j43kljgkld "fds" +,--,fds
#version 330 core
#pragma foo bar fdafds a

===
#custom fdsaflsj 43lk j43kljgkld "fds" +,--,fds
#version 330 core


§testPreprocessor errors,GLSL_PASSTHROUGH,GLSL_CUSTOM_PASSTHROUGH:
#pragma foo bar fdafds a

§testPreprocessor errors,GLSL_CUSTOM_PASSTHROUGH:
#custom fdskalfjk#fdsfldsj

§testPreprocessor errors,GLSL_CUSTOM_PASSTHROUGH:
#version

§testPreprocessor errors,GLSL_PASSTHROUGH:
#custom fds

§testPreprocessor GLSL_PASSTHROUGH,GLSL_CUSTOM_PASSTHROUGH:
#version
#extension
#custom

§testPreprocessor errors,ARBITRARY_PRAGMAS:
#pragma

§testPreprocessor LINE_MARKERS:

===
#line 1 0

§testIncludeLineMapping LINE_MARKERS:
bar
#include "length!2:a"
foo

===
#line 1 0
bar
#line 1 1
content 0
content 1
#line 3 0
foo

§testIncludeLineMapping LINE_MARKERS:
bar
#include "length!2:a"
baz
#include "length!2:a"
foo

===
#line 1 0
bar
#line 1 1
content 0
content 1
#line 3 0
baz
#line 1 1
content 0
content 1
#line 5 0
foo

§testIncludeLineMapping LINE_MARKERS:
bar
#include "length!2:a"
baz
#include "length!2:b"
foo

===
#line 1 0
bar
#line 1 1
content 0
content 1
#line 3 0
baz
#line 1 2
content 0
content 1
#line 5 0
foo

§testIncludeLineMapping LINE_MARKERS,NAMED_LINE_MARKERS:
bar
#include "length!2:a"
foo

===
#line 1 "<no file>"
bar
#line 1 "length!2:a"
content 0
content 1
#line 3 "<no file>"
foo

§testIncludeLineMapping LINE_MARKERS,NAMED_LINE_MARKERS:
bar
#include "length!2:a"
#include "length!2:b"
baz
#include "length!3:c"
foo

===
#line 1 "<no file>"
bar
#line 1 "length!2:a"
content 0
content 1
#line 3 "<no file>"
#line 1 "length!2:b"
content 0
content 1
#line 4 "<no file>"
baz
#line 1 "length!3:c"
content 0
content 1
content 2
#line 6 "<no file>"
foo

§testIncludeLineMappingExplicitFiles LINE_MARKERS,NAMED_LINE_MARKERS:
@b
foo
bar


zoom
@a
a bar
#include "b"
a foo
#include "b"
b zam
@main
bar
#include "a"
foo

===
#line 1 "<no file>"
bar
#line 1 "a"
a bar
#line 1 "b"
foo
bar


zoom
#line 3 "a"
a foo
#line 1 "b"
foo
bar


zoom
#line 5 "a"
b zam
#line 3 "<no file>"
foo

§testIncludeLineMappingExplicitFiles PRAGMA_ONCE,LINE_MARKERS,NAMED_LINE_MARKERS:
@b
foo
bar

#include "a"
zoom
@a
#pragma once
a bar
#include "b"
a foo
#include "b"
b zam
@main
bar
#include "a"
second
#include "a"
foo

===
#line 1 "<no file>"
bar
#line 1 "a"

a bar
#line 1 "b"
foo
bar

#line 1 "a"

#line 5 "b"
zoom
#line 4 "a"
a foo
#line 1 "b"
foo
bar

#line 1 "a"

#line 5 "b"
zoom
#line 6 "a"
b zam
#line 3 "<no file>"
second
#line 1 "a"

#line 5 "<no file>"
foo

§testIncludeLineMappingExplicitFiles GLSL_PASSTHROUGH,LINE_MARKERS,NAMED_LINE_MARKERS:
@a
#version 330 core
#extension GL_ARB_texture_rectangle : enable
bla bla
@b
foo bar
@main
#include "a"
#version 330 core
foo
#include "b"
#extension GL_ARB_texture_rectangle : enable
#include "b"

===
#line 1 "<no file>"
#line 1 "a"
#version 330 core
#extension GL_ARB_texture_rectangle : enable
bla bla
#line 2 "<no file>"
#version 330 core
foo
#line 1 "b"
foo bar
#line 5 "<no file>"
#extension GL_ARB_texture_rectangle : enable
#line 1 "b"
foo bar
#line 7 "<no file>"

§testIncludeLineMappingExplicitFiles LINE_MARKERS,NAMED_LINE_MARKERS:
@a
foo
@main
#include "a"
===
#line 1 "<no file>"
#line 1 "a"
foo

§
