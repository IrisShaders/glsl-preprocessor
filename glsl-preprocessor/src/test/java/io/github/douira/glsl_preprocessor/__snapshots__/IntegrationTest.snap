testIncludeLineMappingExplicitFiles[GLSL_PASSTHROUGH,LINE_MARKERS,NAMED_LINE_MARKERS_Y/Ave1XPFEAn5nYsR2tYDIZlvJU]=[
#include "a"
#version 330 core
foo
#include "b"
#extension GL_ARB_texture_rectangle : enable
#include "b"

<><><><><><><><><><><><><><><><><><><><><><><><><>
(no error)
]


testIncludeLineMappingExplicitFiles[LINE_MARKERS,NAMED_LINE_MARKERS_2x93/pYHABoACwhPVaFOcpcwPI0]=[
bar
#include "a"
foo

<><><><><><><><><><><><><><><><><><><><><><><><><>
(no error)
]


testIncludeLineMappingExplicitFiles[LINE_MARKERS,NAMED_LINE_MARKERS_khv4ydPR8xYuU40++utyw/fXhrE]=[
#include "a"
<><><><><><><><><><><><><><><><><><><><><><><><><>
(no error)
]


testIncludeLineMappingExplicitFiles[PRAGMA_ONCE,LINE_MARKERS,NAMED_LINE_MARKERS_sIEPPAQhrxBRiCEerd7tv3qOzuc]=[
bar
#include "a"
second
#include "a"
foo

<><><><><><><><><><><><><><><><><><><><><><><><><>
(no error)
]


testIncludeLineMapping[LINE_MARKERS,NAMED_LINE_MARKERS_dSPMsT7XryDyTsdqDoJrH6BH3Zo]=[
bar
#include "length!2:a"
foo

<><><><><><><><><><><><><><><><><><><><><><><><><>
(no error)
]


testIncludeLineMapping[LINE_MARKERS,NAMED_LINE_MARKERS_ylY/w6VK6e4/mk6/CiSkpxnOBXs]=[
bar
#include "length!2:a"
#include "length!2:b"
baz
#include "length!3:c"
foo

<><><><><><><><><><><><><><><><><><><><><><><><><>
(no error)
]


testIncludeLineMapping[LINE_MARKERS_dSPMsT7XryDyTsdqDoJrH6BH3Zo]=[
bar
#include "length!2:a"
foo

<><><><><><><><><><><><><><><><><><><><><><><><><>
(no error)
]


testPreprocessor[GLSL_PASSTHROUGH,GLSL_CUSTOM_PASSTHROUGH,ARBITRARY_PRAGMAS_uO+aZa+FgyCVAejH6nxTkTt0CWQ]=[
#custom fdsaflsj 43lk j43kljgkld "fds" +,--,fds
#version 330 core
#pragma foo bar fdafds a

<><><><><><><><><><><><><><><><><><><><><><><><><>
(no error)
]


testPreprocessor[GLSL_PASSTHROUGH,GLSL_CUSTOM_PASSTHROUGH_WD5wCCQXiHxE57nVQkXoycdQjks]=[
#version
#extension
#custom

<><><><><><><><><><><><><><><><><><><><><><><><><>
(no error)
]


testPreprocessor[GLSL_PASSTHROUGH_ORQdau2vuA/2h/xeXX1yjDgBKng]=[
#version 330 core
foo bar
#extension foo : bar
#define DEFINED 4
int a = DEFINED;

<><><><><><><><><><><><><><><><><><><><><><><><><>
(no error)
]


testPreprocessor[GLSL_PASSTHROUGH_bB149WHJMXASfFOJUlrhhdjem24]=[
#line 5 324
#version 330 core
int a = 4;
#line 43 453

foo bar 
#ifdef FOO
FOO is defined here
#endif
#define FOO 5
foo bar
#ifdef FOO
FOO is defined here
#endif

<><><><><><><><><><><><><><><><><><><><><><><><><>
(no error)
]


testPreprocessor[GLSL_PASSTHROUGH_wyy4x0OaDgyEmuM7w5N5SLtVji8]=[
#extension GL_ARB_texture_rectangle : enable
#extension GL_ARB_texture_rectangle : enable
#version 330 fjdsal fjsda fisda jfiojfdsaflkds
#line 1
#if 0
#endif
#line 2
int a = 4;
#define BAR 4
#define FOO 4
if (a == BAR) {
	a = FOO;
}

<><><><><><><><><><><><><><><><><><><><><><><><><>
(no error)
]


testPreprocessor[LINE_MARKERS_2jmj7l5rSw0yVb/vlWAYkK/YBwk]=[

<><><><><><><><><><><><><><><><><><><><><><><><><>
(no error)
]


testPreprocessor[errors,ARBITRARY_PRAGMAS_tkMavAFBAxEG3RtCGFwG7STeCK4]=[
#pragma

<><><><><><><><><><><><><><><><><><><><><><><><><>
io.github.douira.glsl_preprocessor.LexerException: Warning at 1:7: Empty #pragma
]


testPreprocessor[errors,GLSL_CUSTOM_PASSTHROUGH_LiielybpiPYMDUxz8pc85kwvnAI]=[
#custom fdskalfjk#fdsfldsj

<><><><><><><><><><><><><><><><><><><><><><><><><>
io.github.douira.glsl_preprocessor.InternalException: Bad token [#@1,17]:"#"
]


testPreprocessor[errors,GLSL_CUSTOM_PASSTHROUGH_aDv7vwg7JM7S15pnkSb2Vvu1P7U]=[
#version

<><><><><><><><><><><><><><><><><><><><><><><><><>
io.github.douira.glsl_preprocessor.LexerException: GLSL passthrough not enabled
]


testPreprocessor[errors,GLSL_PASSTHROUGH,GLSL_CUSTOM_PASSTHROUGH_SkPEvQvuleWLKW47WWP78/taGeU]=[
#pragma foo bar fdafds a

<><><><><><><><><><><><><><><><><><><><><><><><><>
io.github.douira.glsl_preprocessor.LexerException: Warning at 1:8: Unknown #pragma: foo
]


testPreprocessor[errors,GLSL_PASSTHROUGH_qT/RZ9gQ4GE2a78x7LH4SKT2FWo]=[
#custom fds

<><><><><><><><><><><><><><><><><><><><><><><><><>
io.github.douira.glsl_preprocessor.LexerException: GLSL custom passthrough not enabled
]